src/lamplib/src/genny/tasks/preprocess.py

```diff
@@ -1,42 +1,84 @@
 import os
 import sys
+import tempfile
+import datetime
 from enum import Enum
 from pathlib import Path
 from collections import namedtuple
 from contextlib import AbstractContextManager
 
+from omegaconf import OmegaConf
 import yaml
 import structlog
 
 SLOG = structlog.get_logger(__name__)
+# Cannot be in the default config because yaml merges overwrite lists instead of appending.
 GENNY_INTERNAL = {"Name": "PhaseTimingRecorder", "Type": "PhaseTimingRecorder", "Threads": 1}
+# If this gets any bigger, we should consider a yaml file.
+DEFAULT_CONFIG = {
+    # We go ahead and set the default pool with a 100 size so that
+    # later URI injection can configure it. It's okay if the Default
+    # pool isn't actually used in a workload, since they're constructed
+    # lazily.
+    "Clients": {"Default": {"QueryOptions": {"maxPoolSize": 100}}}
+}
+DEFAULT_URI = "mongodb://localhost:27017"
 
 
 class ParseException(Exception):
     pass
 
 
-def evaluate(workload_path: str, smoke: bool, output: str):
+def evaluate(workload_path: str, smoke: bool, output: str, override_file_path=None):
     """CLI-friendly wrapper for preprocess."""
     if output is not None:
         with open(output, "w") as f:
-            preprocess(workload_path=workload_path, smoke=smoke, output_file=f)
+            preprocess(
+                workload_path=workload_path,
+                smoke=smoke,
+                output_file=f,
+                override_file_path=override_file_path,
+            )
     else:
-        preprocess(workload_path=workload_path, smoke=smoke)
-
-
-def preprocess(workload_path: str, smoke: bool, output_file=None):
+        preprocess(workload_path=workload_path, smoke=smoke, override_file_path=override_file_path)
+
+
+# It's weird to mix our custom preprocessor with OmegaConf.
+# Future work can replace it with OmegaConf resolvers and interpolation.
+def preprocess(
+    workload_path: str,
+    smoke: bool,
+    default_uri: str = DEFAULT_URI,
+    output_file=sys.stdout,
+    override_file_path=None,
+):
     """Evaluate a workload and output it to a file (or stdout)."""
     mode = _ParseMode.Smoke if smoke else _ParseMode.Normal
 
-    parser = _WorkloadParser()
-    output = parser.parse(workload_path, parse_mode=mode)
+    # First, apply the workload yaml over the defaults.
+    conf = OmegaConf.load(workload_path)
+    OmegaConf.create(DEFAULT_CONFIG)
+    conf = OmegaConf.unsafe_merge(DEFAULT_CONFIG, conf)
+
+    # Second, apply any overrides.
+    if override_file_path is not None:
+        overrides = OmegaConf.load(override_file_path)
```

Comment:
That's a really good point. I'm going to capture this and your thoughts in the top-level comment in a separate ticket. My instinct is to move forward as-is but definitely do a more thorough evaluation of Genny's configuration scheme. In particular, I have half-baked ideas of using OmegaConf structured configs for each cpp module. This could let us distribute the config schemas more while type validation lets us do that safely.

The `SchemaVersion` stuff in particular ties into the infra versioning concerns, which we have an (admittedly unstarted) initiative to do and a Genny-specific project for. I especially want to unify Genny and DSI configs. It'd be really nice to have a global `evaluate` before launching a DSI run that would merge in the Genny config too and let you tweak/override things. Will defs keep you in the loop I think this needs a lot more conversation.

Reply:
Created TIG-3717 for this.

Reply:
Ya definitely an interesting idea to start and bike-shed over a lunch or happy hour until it starts to feel real and do-able. I'll look forward to that.